File: web/rgbd-js/integer_frame.js
import { NativeInt32Array } from "./capi_containers.js";

export class Int32Frame {
  constructor(nativeInt32Frame) {
    this.width = nativeInt32Frame.getWidth();
    this.height = nativeInt32Frame.getHeight();
    this.values = nativeInt32Frame.getValues();
  }
}

export class NativeInt32Frame {
  constructor(wasmModule, ptr) {
    this.wasmModule = wasmModule;
    this.ptr = ptr;
  }

  close() {
    this.wasmModule.ccall("rgbd_int32_frame_dtor", null, ["number"], [this.ptr]);
  }

  getWidth() {
    return this.wasmModule.ccall("rgbd_int32_frame_get_width", "number", ["number"], [this.ptr]);
  }

  getHeight() {
    return this.wasmModule.ccall("rgbd_int32_frame_get_height", "number", ["number"], [this.ptr]);
  }

  getValues() {
    const nativeValuesPtr = this.wasmModule.ccall("rgbd_int32_frame_get_values", "number", ["number"], [this.ptr]);
    const nativeValues = new NativeInt32Array(this.wasmModule, nativeValuesPtr);
    const values = nativeValues.toArray();
    nativeValues.close();
    return values;
  }
}

File: python/pyrgbd/integer_frame.py
import numpy as np
from ._librgbd_ffi import lib
from .capi_containers import NativeInt32Array
from .utils import cast_np_array_to_pointer


class NativeInt32Frame:
    def __init__(self, ptr):
        self.ptr = ptr

    def close(self):
        lib.rgbd_int32_frame_dtor(self.ptr)

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.close()

    def get_width(self) -> int:
        return lib.rgbd_int32_frame_get_width(self.ptr)

    def get_height(self) -> int:
        return lib.rgbd_int32_frame_get_height(self.ptr)

    def get_values(self) -> NativeInt32Array:
        return NativeInt32Array(lib.rgbd_int32_frame_get_values(self.ptr))


class Int32Frame:
    def __init__(self, width: int, height: int, values: np.ndarray):
        self.width = width
        self.height = height
        self.values = values

    @classmethod
    def from_native(cls, native_int32_frame: NativeInt32Frame):
        width = native_int32_frame.get_width()
        height = native_int32_frame.get_height()
        values = native_int32_frame.get_values().to_np_array().reshape((height, width))
        return Int32Frame(width, height, values)

    def to_native(self):
        ptr = lib.rgbd_int32_frame_ctor(
            self.width, self.height, cast_np_array_to_pointer(self.values)
        )
        return NativeInt32Frame(ptr)
